### DISCUSSION TITLE
Styling HTML elements

### DISCUSSION THREAD

**Discussion by jkjustjoshing** [jpudysz/react-native-unistyles#383] [Q&A]
I have a React Native project where our only build output targets the web and `react-native-web`. We don't ship any native apps at the moment. We're trying to decide if we want to adopt Unistyles.

We have a few web-only components where we're styling HTML elements using the `className` prop, i.e. `<div>`, `<p>`. How does it work to add class names to these elements with Unistyles? Is the class name exposed at runtime to the component in the `styles` object? Is there a mechanism for this?

**Comment by jpudysz** [comment]
Hey!  
We're connecting classes based on an element's ref, similar to native ShadowNodes.  
We're also managing the lifecycle of these linkages.  

If you have HTML elements like `<div>` or `<p>`, why not add such classes inline? Unistyles handles adding classes to refs based on `StyleSheet` styles. You must know that there is no `className` for native, so RNW fixes that. 

It would be nice if you could give me winder context of what exactly you need and how you want to use these class names.

**Reply by jkjustjoshing** [reply]
Here's a basic example:

```ts
// MyComponent.web.ts

import { StyleSheet } from 'react-native-unistyles'

const MyComponent = () => {
  return (
    <div className={/* what do I put here to apply a red background color? */} />
  )
}

const styles = StyleSheet.create({
   container: {
     backgroundColor: 'red'
   }
})
```

**Reply by jpudysz** [reply]
Ah I see! 

This **could** work. With beta.1 it won't though because Babel plugin supports only RN components. I can add config to extend it with extra component names. 

One catch: you need to pass styles.xyz to style prop not className. But it will work as you expect. 

Let me know if you want to support such case. We can add it to beta.2

**Reply by jkjustjoshing** [reply]
Without this functionality we wouldn't be able to adopt Unistyles into our repo, so we would love to see this!

**Reply by jkjustjoshing** [reply]
And to clarify, if you supported this and we passed `styles.xyz` to the `style` prop of the HTML element, would the rendered output HTML element have inline styles applied, or styles applied with a class name?

**Reply by jpudysz** [reply]
No inline styles. It will always be class + className on div. We're React Native first, so there is no support for direct `className` prop**. 

You can test it in `beta.4` with React Native + React Native Web. Open your example in Chrome and check the classes. Everything that unistyles outputs has a `unistyle` prefix in the class name.

** - you can also add className from StyleSheet: https://www.unistyl.es/v3/references/web-only#injecting-custom-classnames

**Reply by jkjustjoshing** [reply]
I just tried this with `beta.4`:

```tsx
import { StyleSheet } from "react-native-unistyles";

import { PropsWithChildren } from "react";

export function Text({ children }: PropsWithChildren) {
  return <p style={styles.text}>{children}</p>;
}

const styles = StyleSheet.create({
  text: {
    fontSize: 16,
    lineHeight: 24,
    _web: {
      backgroundColor: "purple",
    },
  },
});
```

When I render that component, I get zero styles on the `<p>` element (no `style` or `class` attributes). When I build the project for web, this is what the HTML elements look like
```
<p style="__uni__secrets__:[object Object]">Rendered text</p>
```

Am I missing anything, or is the Unistyles implementation for this use case still incomplete?

**Reply by jpudysz** [reply]
You need to use View or Text from RN. We don’t suport HTML tags yet

**Reply by jkjustjoshing** [reply]
But using HTML tags is the whole point of this discussion.

**Reply by jpudysz** [reply]
I didn’t say it works now. I said that it’s possible (first reply). You asked about inline styles vs class names. So with beta4 we don’t do inline styles, you still need to use React Native components.

**Reply by jkjustjoshing** [reply]
Hmm, then I'm a bit confused, and/or we're not 100% understanding each other.

> Open your example in Chrome and check the classes

My example involved setting styles on a `<div>` JSX element, so when you said that I created a project with `beta.4` and tried adding styles to the `<div>` (as per my example). You said in your first response that it's not supported by `beta.1`, but then you said I should test it in `beta.4`, so it sounded like somewhere in betas 2-4 you added the ability to add styles to a `<div>` (which, again, is the thing I'm trying to do).

> You asked about inline styles vs class names

I [asked that](https://github.com/jpudysz/react-native-unistyles/discussions/383#discussioncomment-11582817) in the context of adding styles to a `<div>` in the JSX.

**Reply by jpudysz** [reply]
It was about your question about inline styles. But nevermind, it's still not supported.

**Reply by jkjustjoshing** [reply]
I understand.

Is this something you'd consider supporting in the future?

**Reply by jpudysz** [reply]
I do, but no plans for the next betas. Maybe closer to RC.

**Reply by remilry** [reply]
Hi, is there any plan for this now ?

**Reply by jpudysz** [reply]
https://www.unistyl.es/v3/guides/custom-web

**Reply by remilry** [reply]
Oh I missed that, thank you 😄
